File: src/common/Jarvis.PiAgentHost/src/desktop-model-broker.mjs
import {
  createAssistantMessageEventStream,
} from "@earendil-works/pi-ai";
import { randomUUID } from "node:crypto";
import { once } from "node:events";
import { createConnection } from "node:net";

const brokerProtocol = "jarvisv2-pi-model-broker-v1";
const brokerProviderId = "jarvis-desktop-broker";
const brokerModelId = "desktop-default";
const maximumBrokerFrameBytes = 1_048_576;
const pipePattern =
  /^\\\\\.\\pipe\\jarvis2-pi-model-[0-9a-f]{32}$/iu;

function emptyUsage() {
  return {
    input: 0,
    output: 0,
    cacheRead: 0,
    cacheWrite: 0,
    totalTokens: 0,
    cost: {
      input: 0,
      output: 0,
      cacheRead: 0,
      cacheWrite: 0,
      total: 0,
    },
  };
}

function createOutput(model) {
  return {
    role: "assistant",
    content: [],
    api: model.api,
    provider: model.provider,
    model: model.id,
    usage: emptyUsage(),
    stopReason: "stop",
    timestamp: Date.now(),
  };
}

function boundedInteger(value) {
  return Number.isSafeInteger(value) && value >= 0
    ? value
    : 0;
}

function applyUsage(output, usage) {
  output.usage.input = boundedInteger(usage?.input);
  output.usage.output = boundedInteger(usage?.output);
  output.usage.cacheRead = boundedInteger(usage?.cacheRead);
  output.usage.cacheWrite = boundedInteger(usage?.cacheWrite);
  output.usage.totalTokens =
    output.usage.input +
    output.usage.output +
    output.usage.cacheRead +
    output.usage.cacheWrite;
}

async function writeRecord(socket, record) {
  const line = JSON.stringify(record);
  if (Buffer.byteLength(line, "utf8") > maximumBrokerFrameBytes) {
    throw new Error("Desktop model broker request exceeded its frame limit.");
  }
  if (!socket.write(`${line}\n`, "utf8")) {
    await once(socket, "drain");
  }
}

async function* readRecords(socket) {
  socket.setEncoding("utf8");
  let buffer = "";
  for await (const chunk of socket) {
    buffer += chunk;
    let newlineIndex = buffer.indexOf("\n");
    while (newlineIndex >= 0) {
      const line = buffer.slice(0, newlineIndex);
      buffer = buffer.slice(newlineIndex + 1);
      newlineIndex = buffer.indexOf("\n");
      if (line.length === 0) {
        continue;
      }
      if (
        Buffer.byteLength(line, "utf8") >
        maximumBrokerFrameBytes
      ) {
        throw new Error(
          "Desktop model broker response exceeded its frame limit.",
        );
      }
      yield JSON.parse(line);
    }
    if (
      Buffer.byteLength(buffer, "utf8") >
      maximumBrokerFrameBytes
    ) {
      throw new Error(
        "Desktop model broker pending frame exceeded its limit.",
      );
    }
  }
  if (buffer.trim().length !== 0) {
    throw new Error(
      "Desktop model broker closed with an incomplete frame.",
    );
  }
}

async function connectBroker(pipePath, signal) {
  const socket = createConnection(pipePath);
  const abort = () => {
    socket.destroy(new Error("Desktop model request aborted."));
  };
  signal?.addEventListener("abort", abort, { once: true });
  try {
    await new Promise((resolve, reject) => {
      const connected = () => {
        socket.off("error", failed);
        resolve();
      };
      const failed = (error) => {
        socket.off("connect", connected);
        reject(error);
      };
      socket.once("connect", connected);
      socket.once("error", failed);
    });
  } catch (error) {
    signal?.removeEventListener("abort", abort);
    socket.destroy();
    throw error;
  }
  return {
    socket,
    detachAbort: () =>
      signal?.removeEventListener("abort", abort),
  };
}

function createBrokerStream(pipePath) {
  return (model, context, options) => {
    const stream = createAssistantMessageEventStream();
    const output = createOutput(model);

    (async () => {
      let socket;
      let detachAbort = () => {};
      try {
        const connection = await connectBroker(
          pipePath,
          options?.signal,
        );
        socket = connection.socket;
        detachAbort = connection.detachAbort;
        const requestId = randomUUID();
        const records = readRecords(socket);
        await writeRecord(socket, {
          type: "broker_hello",
          protocol: brokerProtocol,
        });
        const ready = await records.next();
        if (
          ready.done ||
          ready.value?.type !== "broker_ready" ||
          ready.value?.protocol !== brokerProtocol
        ) {
          throw new Error(
            "Desktop model broker failed protocol admission.",
          );
        }

        await writeRecord(socket, {
          type: "model_request",
          protocol: brokerProtocol,
          id: requestId,
          model: {
            provider: model.provider,
            id: model.id,
          },
          context,
          options: {
            maxTokens: options?.maxTokens,
            reasoning: options?.reasoning,
            sessionId: options?.sessionId,
          },
        });

        stream.push({ type: "start", partial: output });
        const textBlock = {
          type: "text",
          text: "",
        };
        output.content.push(textBlock);
        stream.push({
          type: "text_start",
          contentIndex: 0,
          partial: output,
        });
        const toolCalls = new Map();

        for await (const record of records) {
          if (record?.id !== requestId) {
            throw new Error(
              "Desktop model broker response id did not match.",
            );
          }
          if (record.type === "model_delta") {
            if (typeof record.delta !== "string") {
              throw new Error(
                "Desktop model broker emitted an invalid text delta.",
              );
            }
            textBlock.text += record.delta;
            stream.push({
              type: "text_delta",
              contentIndex: 0,
              delta: record.delta,
              partial: output,
            });
            continue;
          }
          if (record.type === "model_tool_call_start") {
            if (
              typeof record.toolCallId !== "string" ||
              typeof record.name !== "string" ||
              toolCalls.has(record.toolCallId)
            ) {
              throw new Error(
                "Desktop model broker emitted an invalid tool start.",
              );
            }
            const toolCall = {
              type: "toolCall",
              id: record.toolCallId,
              name: record.name,
              arguments: {},
              partialJson: "",
            };
            const contentIndex = output.content.length;
            output.content.push(toolCall);
            toolCalls.set(record.toolCallId, {
              contentIndex,
              toolCall,
            });
            stream.push({
              type: "toolcall_start",
              contentIndex,
              partial: output,
            });
            continue;
          }
          if (record.type === "model_tool_call_delta") {
            const state = toolCalls.get(record.toolCallId);
            if (!state || typeof record.delta !== "string") {
              throw new Error(
                "Desktop model broker emitted an invalid tool delta.",
              );
            }
            state.toolCall.partialJson += record.delta;
            try {
              state.toolCall.arguments = JSON.parse(
                state.toolCall.partialJson,
              );
            } catch {
              // Partial JSON is expected while a tool call is streaming.
            }
            stream.push({
              type: "toolcall_delta",
              contentIndex: state.contentIndex,
              delta: record.delta,
              partial: output,
            });
            continue;
          }
          if (record.type === "model_tool_call_end") {
            const state = toolCalls.get(record.toolCallId);
            if (!state) {
              throw new Error(
                "Desktop model broker ended an unknown tool call.",
              );
            }
            state.toolCall.arguments = JSON.parse(
              state.toolCall.partialJson || "{}",
            );
            delete state.toolCall.partialJson;
            stream.push({
              type: "toolcall_end",
              contentIndex: state.contentIndex,
              toolCall: state.toolCall,
              partial: output,
            });
            toolCalls.delete(record.toolCallId);
            continue;
          }
          if (record.type === "model_error") {
            throw new Error(
              typeof record.message === "string"
                ? record.message
                : "Desktop model broker rejected the request.",
            );
          }
          if (record.type === "model_done") {
            if (toolCalls.size !== 0) {
              throw new Error(
                "Desktop model broker left a tool call incomplete.",
              );
            }
            stream.push({
              type: "text_end",
              contentIndex: 0,
              content: textBlock.text,
              partial: output,
            });
            output.stopReason = [
              "stop",
              "length",
              "toolUse",
            ].includes(record.reason)
              ? record.reason
              : "stop";
            applyUsage(output, record.usage);
            stream.push({
              type: "done",
              reason: output.stopReason,
              message: output,
            });
            stream.end();
            socket.end();
            return;
          }
          throw new Error(
            "Desktop model broker emitted an unsupported frame.",
          );
        }
        throw new Error(
          "Desktop model broker closed before completing the request.",
        );
      } catch (error) {
        output.stopReason = options?.signal?.aborted
          ? "aborted"
          : "error";
        output.errorMessage =
          error instanceof Error
            ? error.message
            : String(error);
        stream.push({
          type: "error",
          reason: output.stopReason,
          error: output,
        });
        stream.end();
        socket?.destroy();
      } finally {
        detachAbort();
      }
    })();

    return stream;
  };
}

export function validateDesktopBrokerPipe(pipePath) {
  return (
    typeof pipePath === "string" &&
    pipePattern.test(pipePath)
  );
}

export function registerDesktopBrokerProvider(
  modelRuntime,
  pipePath,
) {
  if (!validateDesktopBrokerPipe(pipePath)) {
    throw new Error(
      "Desktop model broker pipe failed local-path admission.",
    );
  }
  modelRuntime.registerProvider(brokerProviderId, {
    name: "JARVIS Desktop Model Broker",
    api: "jarvis-desktop-broker",
    apiKey: "desktop-broker-capability",
    streamSimple: createBrokerStream(pipePath),
    models: [{
      id: brokerModelId,
      name: "JARVIS Desktop Default",
      api: "jarvis-desktop-broker",
      baseUrl: "jarvis-desktop-pipe",
      reasoning: false,
      input: ["text"],
      cost: {
        input: 0,
        output: 0,
        cacheRead: 0,
        cacheWrite: 0,
      },
      contextWindow: 128_000,
      maxTokens: 16_384,
    }],
  });
  const model = modelRuntime.getModel(
    brokerProviderId,
    brokerModelId,
  );
  if (!model || !modelRuntime.hasConfiguredAuth(brokerProviderId)) {
    throw new Error(
      "Desktop model broker provider failed model admission.",
    );
  }
  return model;
}

export {
  brokerModelId,
  brokerProtocol,
  brokerProviderId,
  maximumBrokerFrameBytes,
};

File: src/common/Jarvis.PiAgentHost/DesktopModelBroker.cs
using System.Collections.Concurrent;
using System.IO.Pipes;
using System.Runtime.CompilerServices;
using System.Text;
using System.Text.Json;

namespace Jarvis.PiAgentHost;

public sealed record DesktopModelBrokerRequest(
    string RequestId,
    string ProviderId,
    string ModelId,
    JsonElement Context,
    JsonElement Options);

public sealed record DesktopModelUsage(
    long Input,
    long Output,
    long CacheRead,
    long CacheWrite);

public abstract record DesktopModelStreamEvent;

public sealed record DesktopModelTextDelta(
    string Delta) : DesktopModelStreamEvent;

public sealed record DesktopModelToolCallStarted(
    string ToolCallId,
    string Name) : DesktopModelStreamEvent;

public sealed record DesktopModelToolCallDelta(
    string ToolCallId,
    string Delta) : DesktopModelStreamEvent;

public sealed record DesktopModelToolCallCompleted(
    string ToolCallId) : DesktopModelStreamEvent;

public sealed record DesktopModelCompleted(
    string Reason,
    DesktopModelUsage Usage) : DesktopModelStreamEvent;

public interface IDesktopModelProvider
{
    IAsyncEnumerable<DesktopModelStreamEvent> StreamAsync(
        DesktopModelBrokerRequest request,
        CancellationToken cancellationToken);
}

public sealed class DesktopModelBrokerServer : IAsyncDisposable
{
    public const string Protocol = "jarvisv2-pi-model-broker-v1";
    public const string ProviderId = "jarvis-desktop-broker";
    public const string ModelId = "desktop-default";
    public const int MaximumFrameBytes = 1_048_576;
    public const int MaximumConcurrentConnections = 4;

    private static readonly IReadOnlySet<string> AllowedToolNames =
        new HashSet<string>(
            ["read", "grep", "find", "ls"],
            StringComparer.Ordinal);

    private static readonly HashSet<string> ForbiddenCredentialFields =
        new(StringComparer.OrdinalIgnoreCase)
        {
            "apikey",
            "authorization",
            "credential",
            "credentials",
            "password",
            "secret",
            "token",
            "accesstoken",
            "refreshtoken",
        };

    private readonly CancellationTokenSource cancellation = new();
    private readonly ConcurrentDictionary<int, Task> clients = new();
    private readonly SemaphoreSlim clientSlots = new(
        MaximumConcurrentConnections,
        MaximumConcurrentConnections);
    private readonly IDesktopModelProvider provider;
    private readonly Task acceptTask;
    private int clientSequence;
    private int faultCount;
    private int requestCount;
    private NamedPipeServerStream? pendingAcceptPipe;

    public string PipePath { get; }
    public int FaultCount => Volatile.Read(ref faultCount);
    public int RequestCount => Volatile.Read(ref requestCount);

    private DesktopModelBrokerServer(
        IDesktopModelProvider provider)
    {
        this.provider = provider;
        string pipeName = $"jarvis2-pi-model-{Guid.NewGuid():N}";
        PipePath = $@"\\.\pipe\{pipeName}";
        acceptTask = AcceptClientsAsync(
            pipeName,
            cancellation.Token);
    }

    public static DesktopModelBrokerServer Start(
        IDesktopModelProvider provider)
    {
        ArgumentNullException.ThrowIfNull(provider);
        return new DesktopModelBrokerServer(provider);
    }

    private async Task AcceptClientsAsync(
        string pipeName,
        CancellationToken cancellationToken)
    {
        while (!cancellationToken.IsCancellationRequested)
        {
            try
            {
                await clientSlots.WaitAsync(cancellationToken);
            }
            catch (OperationCanceledException)
                when (cancellationToken.IsCancellationRequested)
            {
                return;
            }
            if (cancellationToken.IsCancellationRequested)
            {
                clientSlots.Release();
                return;
            }

            NamedPipeServerStream pipe;
            try
            {
                pipe = CreatePipe(pipeName);
            }
            catch
            {
                clientSlots.Release();
                throw;
            }
            Volatile.Write(ref pendingAcceptPipe, pipe);
            try
            {
                await pipe.WaitForConnectionAsync(cancellationToken);
            }
            catch (OperationCanceledException)
                when (cancellationToken.IsCancellationRequested)
            {
                await pipe.DisposeAsync();
                clientSlots.Release();
                return;
            }
            catch (ObjectDisposedException)
                when (cancellationToken.IsCancellationRequested)
            {
                clientSlots.Release();
                return;
            }
            catch (IOException)
                when (cancellationToken.IsCancellationRequested)
            {
                await pipe.DisposeAsync();
                clientSlots.Release();
                return;
            }
            catch
            {
                await pipe.DisposeAsync();
                clientSlots.Release();
                throw;
            }
            finally
            {
                Interlocked.CompareExchange(
                    ref pendingAcceptPipe,
                    null,
                    pipe);
            }

            if (cancellationToken.IsCancellationRequested)
            {
                await pipe.DisposeAsync();
                clientSlots.Release();
                return;
            }

            int clientId = Interlocked.Increment(
                ref clientSequence);
            Task clientTask = HandleClientSafelyAsync(
                pipe,
                cancellationToken);
            if (!clients.TryAdd(clientId, clientTask))
            {
                await pipe.DisposeAsync();
                clientSlots.Release();
                throw new InvalidOperationException(
                    "The model broker client id collided.");
            }
            _ = RemoveCompletedClientAsync(
                clientId,
                clientTask);
        }
    }

    private static NamedPipeServerStream CreatePipe(
        string pipeName)
    {
        return new NamedPipeServerStream(
            pipeName,
            PipeDirection.InOut,
            MaximumConcurrentConnections,
            PipeTransmissionMode.Byte,
            PipeOptions.Asynchronous | PipeOptions.CurrentUserOnly,
            8_192,
            8_192);
    }

    private async Task RemoveCompletedClientAsync(
        int clientId,
        Task clientTask)
    {
        try
        {
            await clientTask;
        }
        finally
        {
            clients.TryRemove(clientId, out _);
            clientSlots.Release();
        }
    }

    private async Task HandleClientSafelyAsync(
        NamedPipeServerStream pipe,
        CancellationToken cancellationToken)
    {
        await using (pipe)
        {
            try
            {
                await HandleClientAsync(pipe, cancellationToken);
            }
            catch (OperationCanceledException)
                when (cancellationToken.IsCancellationRequested)
            {
            }
            catch (IOException)
            {
                Interlocked.Increment(ref faultCount);
            }
            catch
            {
                Interlocked.Increment(ref faultCount);
            }
        }
    }

    private async Task HandleClientAsync(
        NamedPipeServerStream pipe,
        CancellationToken cancellationToken)
    {
        using StreamReader reader = new(
            pipe,
            new UTF8Encoding(false, true),
            detectEncodingFromByteOrderMarks: false,
            bufferSize: 8_192,
            leaveOpen: true);
        using StreamWriter writer = new(
            pipe,
            new UTF8Encoding(false, true),
            bufferSize: 8_192,
            leaveOpen: true)
        {
            AutoFlush = true,
            NewLine = "\n",
        };

        using JsonDocument hello = await ReadFrameAsync(
            reader,
            cancellationToken);
        JsonElement helloRoot = hello.RootElement;
        if (
            helloRoot.GetProperty("type").GetString() != "broker_hello" ||
            helloRoot.GetProperty("protocol").GetString() != Protocol)
        {
            throw new InvalidOperationException(
                "The model broker hello failed admission.");
        }
        await WriteFrameAsync(
            writer,
            new
            {
                type = "broker_ready",
                protocol = Protocol,
            },
            cancellationToken);

        using JsonDocument requestFrame = await ReadFrameAsync(
            reader,
            cancellationToken);
        DesktopModelBrokerRequest request = ParseRequest(
            requestFrame.RootElement);
        Interlocked.Increment(ref requestCount);

        using CancellationTokenSource providerCancellation =
            CancellationTokenSource.CreateLinkedTokenSource(
                cancellationToken);
        using CancellationTokenSource monitorCancellation =
            CancellationTokenSource.CreateLinkedTokenSource(
                cancellationToken);
        Task disconnectMonitor = MonitorDisconnectAsync(
            reader,
            providerCancellation,
            monitorCancellation.Token);

        bool completed = false;
        HashSet<string> activeToolCalls =
            new(StringComparer.Ordinal);
        try
        {
            await foreach (
                DesktopModelStreamEvent streamEvent in provider
                    .StreamAsync(
                        request,
                        providerCancellation.Token)
                    .WithCancellation(providerCancellation.Token))
            {
                if (completed)
                {
                    throw new InvalidOperationException(
                        "The model provider emitted data after completion.");
                }
                completed = await WriteProviderEventAsync(
                    writer,
                    request.RequestId,
                    streamEvent,
                    activeToolCalls,
                    providerCancellation.Token);
            }
            if (!completed && !providerCancellation.IsCancellationRequested)
            {
                throw new InvalidOperationException(
                    "The model provider ended without completion.");
            }
        }
        catch (OperationCanceledException)
            when (providerCancellation.IsCancellationRequested)
        {
        }
        catch
        {
            Interlocked.Increment(ref faultCount);
            await TryWriteProviderErrorAsync(
                writer,
                request.RequestId,
                cancellationToken);
        }
        finally
        {
            monitorCancellation.Cancel();
            try
            {
                await disconnectMonitor;
            }
            catch (OperationCanceledException)
            {
            }
            catch (IOException)
            {
            }
        }
    }

    private static DesktopModelBrokerRequest ParseRequest(
        JsonElement root)
    {
        if (
            root.GetProperty("type").GetString() != "model_request" ||
            root.GetProperty("protocol").GetString() != Protocol ||
            ContainsCredentialField(root))
        {
            throw new InvalidOperationException(
                "The model request failed desktop admission.");
        }
        string requestId =
            root.GetProperty("id").GetString()
            ?? throw new InvalidOperationException(
                "The model request id was missing.");
        if (
            !Guid.TryParseExact(requestId, "D", out _) ||
            requestId.Length != 36)
        {
            throw new InvalidOperationException(
                "The model request id was invalid.");
        }

        JsonElement model = root.GetProperty("model");
        string providerId =
            model.GetProperty("provider").GetString()
            ?? string.Empty;
        string modelId =
            model.GetProperty("id").GetString()
            ?? string.Empty;
        if (providerId != ProviderId || modelId != ModelId)
        {
            throw new InvalidOperationException(
                "The model request target was not admitted.");
        }
        JsonElement context = root.GetProperty("context");
        JsonElement options = root.GetProperty("options");
        if (
            context.ValueKind != JsonValueKind.Object ||
            options.ValueKind != JsonValueKind.Object)
        {
            throw new InvalidOperationException(
                "The model request payload shape was invalid.");
        }
        return new DesktopModelBrokerRequest(
            requestId,
            providerId,
            modelId,
            context.Clone(),
            options.Clone());
    }

    private static async Task MonitorDisconnectAsync(
        StreamReader reader,
        CancellationTokenSource providerCancellation,
        CancellationToken cancellationToken)
    {
        try
        {
            string? unexpected = await reader.ReadLineAsync(
                cancellationToken);
            providerCancellation.Cancel();
            if (unexpected is not null)
            {
                throw new InvalidOperationException(
                    "The model broker peer sent an unexpected frame.");
            }
        }
        catch (OperationCanceledException)
            when (cancellationToken.IsCancellationRequested)
        {
        }
    }

    private static async Task<bool> WriteProviderEventAsync(
        StreamWriter writer,
        string requestId,
        DesktopModelStreamEvent streamEvent,
        ISet<string> activeToolCalls,
        CancellationToken cancellationToken)
    {
        switch (streamEvent)
        {
            case DesktopModelTextDelta text:
                if (text.Delta.Length == 0)
                {
                    return false;
                }
                await WriteFrameAsync(
                    writer,
                    new
                    {
                        type = "model_delta",
                        id = requestId,
                        delta = text.Delta,
                    },
                    cancellationToken);
                return false;
            case DesktopModelToolCallStarted toolStart:
                ValidateToolIdentity(
                    toolStart.ToolCallId,
                    toolStart.Name);
                if (
                    !AllowedToolNames.Contains(toolStart.Name) ||
                    !activeToolCalls.Add(toolStart.ToolCallId))
                {
                    throw new InvalidOperationException(
                        "The model provider emitted an unadmitted tool start.");
                }
                await WriteFrameAsync(
                    writer,
                    new
                    {
                        type = "model_tool_call_start",
                        id = requestId,
                        toolCallId = toolStart.ToolCallId,
                        name = toolStart.Name,
                    },
                    cancellationToken);
                return false;
            case DesktopModelToolCallDelta toolDelta:
                ValidateToolIdentity(
                    toolDelta.ToolCallId,
                    name: null);
                if (!activeToolCalls.Contains(toolDelta.ToolCallId))
                {
                    throw new InvalidOperationException(
                        "The model provider emitted a tool delta without a start.");
                }
                await WriteFrameAsync(
                    writer,
                    new
                    {
                        type = "model_tool_call_delta",
                        id = requestId,
                        toolCallId = toolDelta.ToolCallId,
                        delta = toolDelta.Delta,
                    },
                    cancellationToken);
                return false;
            case DesktopModelToolCallCompleted toolEnd:
                ValidateToolIdentity(
                    toolEnd.ToolCallId,
                    name: null);
                if (!activeToolCalls.Remove(toolEnd.ToolCallId))
                {
                    throw new InvalidOperationException(
                        "The model provider ended an inactive tool call.");
                }
                await WriteFrameAsync(
                    writer,
                    new
                    {
                        type = "model_tool_call_end",
                        id = requestId,
                        toolCallId = toolEnd.ToolCallId,
                    },
                    cancellationToken);
                return false;
            case DesktopModelCompleted completed:
                if (activeToolCalls.Count != 0)
                {
                    throw new InvalidOperationException(
                        "The model provider completed with active tool calls.");
                }
                string reason = completed.Reason is
                    "stop" or "length" or "toolUse"
                    ? completed.Reason
                    : throw new InvalidOperationException(
                        "The model completion reason was invalid.");
                ValidateUsage(completed.Usage);
                await WriteFrameAsync(
                    writer,
                    new
                    {
                        type = "model_done",
                        id = requestId,
                        reason,
                        usage = new
                        {
                            input = completed.Usage.Input,
                            output = completed.Usage.Output,
                            cacheRead = completed.Usage.CacheRead,
                            cacheWrite = completed.Usage.CacheWrite,
                        },
                    },
                    cancellationToken);
                return true;
            default:
                throw new InvalidOperationException(
                    "The model provider event was not admitted.");
        }
    }

    private static void ValidateToolIdentity(
        string toolCallId,
        string? name)
    {
        if (
            string.IsNullOrWhiteSpace(toolCallId) ||
            toolCallId.Length > 256 ||
            (name is not null &&
                (string.IsNullOrWhiteSpace(name) ||
                    name.Length > 128)))
        {
            throw new InvalidOperationException(
                "The model tool identity was invalid.");
        }
    }

    private static void ValidateUsage(DesktopModelUsage usage)
    {
        if (
            usage.Input < 0 ||
            usage.Output < 0 ||
            usage.CacheRead < 0 ||
            usage.CacheWrite < 0)
        {
            throw new InvalidOperationException(
                "The model usage values were invalid.");
        }
    }

    private static async Task<JsonDocument> ReadFrameAsync(
        StreamReader reader,
        CancellationToken cancellationToken)
    {
        string? line = await reader.ReadLineAsync(cancellationToken);
        if (line is null)
        {
            throw new IOException(
                "The model broker peer closed before a complete frame.");
        }
        if (Encoding.UTF8.GetByteCount(line) > MaximumFrameBytes)
        {
            throw new InvalidOperationException(
                "The model broker input exceeded its frame limit.");
        }
        try
        {
            return JsonDocument.Parse(line);
        }
        catch (JsonException exception)
        {
            throw new InvalidOperationException(
                "The model broker input was invalid JSON.",
                exception);
        }
    }

    private static async Task WriteFrameAsync<T>(
        StreamWriter writer,
        T value,
        CancellationToken cancellationToken)
    {
        string payload = JsonSerializer.Serialize(value);
        if (Encoding.UTF8.GetByteCount(payload) > MaximumFrameBytes)
        {
            throw new InvalidOperationException(
                "The model broker output exceeded its frame limit.");
        }
        await writer.WriteLineAsync(
            payload.AsMemory(),
            cancellationToken);
    }

    private static async Task TryWriteProviderErrorAsync(
        StreamWriter writer,
        string requestId,
        CancellationToken cancellationToken)
    {
        try
        {
            await WriteFrameAsync(
                writer,
                new
                {
                    type = "model_error",
                    id = requestId,
                    message =
                        "The desktop model provider failed closed.",
                },
                cancellationToken);
        }
        catch
        {
        }
    }

    private static bool ContainsCredentialField(JsonElement element)
    {
        if (element.ValueKind == JsonValueKind.Array)
        {
            return element
                .EnumerateArray()
                .Any(ContainsCredentialField);
        }
        if (element.ValueKind != JsonValueKind.Object)
        {
            return false;
        }
        foreach (JsonProperty property in element.EnumerateObject())
        {
            string normalized = property.Name
                .Replace("-", string.Empty, StringComparison.Ordinal)
                .Replace("_", string.Empty, StringComparison.Ordinal);
            if (
                ForbiddenCredentialFields.Contains(normalized) ||
                ContainsCredentialField(property.Value))
            {
                return true;
            }
        }
        return false;
    }

    public async ValueTask DisposeAsync()
    {
        cancellation.Cancel();
        NamedPipeServerStream? pendingPipe =
            Interlocked.Exchange(
                ref pendingAcceptPipe,
                null);
        if (pendingPipe is not null)
        {
            await pendingPipe.DisposeAsync();
        }
        try
        {
            await acceptTask;
        }
        catch (OperationCanceledException)
        {
        }
        catch (ObjectDisposedException)
        {
        }

        Task[] activeClients = clients.Values.ToArray();
        if (activeClients.Length != 0)
        {
            await Task.WhenAll(activeClients);
        }
        if (provider is IAsyncDisposable asyncDisposable)
        {
            await asyncDisposable.DisposeAsync();
        }
        else if (provider is IDisposable disposable)
        {
            disposable.Dispose();
        }
        cancellation.Dispose();
        clientSlots.Dispose();
    }
}

internal sealed class DiagnosticDesktopModelProvider :
    IDesktopModelProvider
{
    private readonly bool holdResponse;
    private readonly TaskCompletionSource<bool> requestObserved = new(
        TaskCreationOptions.RunContinuationsAsynchronously);
    private readonly ConcurrentQueue<JsonElement> requestContexts = new();
    private int requestSequence;

    public DiagnosticDesktopModelProvider(bool holdResponse)
    {
        this.holdResponse = holdResponse;
    }

    public IReadOnlyList<JsonElement> RequestContexts =>
        requestContexts.ToArray();

    public async Task WaitForRequestAsync(
        CancellationToken cancellationToken)
    {
        await requestObserved.Task.WaitAsync(cancellationToken);
    }

    public async IAsyncEnumerable<DesktopModelStreamEvent> StreamAsync(
        DesktopModelBrokerRequest request,
        [EnumeratorCancellation]
        CancellationToken cancellationToken)
    {
        requestObserved.TrySetResult(true);
        requestContexts.Enqueue(request.Context.Clone());
        int sequence = Interlocked.Increment(
            ref requestSequence);
        if (holdResponse)
        {
            await Task.Delay(
                Timeout.InfiniteTimeSpan,
                cancellationToken);
            yield break;
        }

        if (sequence == 3)
        {
            yield return new DesktopModelToolCallStarted(
                "diagnostic-read-1",
                "read");
            yield return new DesktopModelToolCallDelta(
                "diagnostic-read-1",
                "{\"path\":\"package.json\",\"offset\":1,\"limit\":5}");
            yield return new DesktopModelToolCallCompleted(
                "diagnostic-read-1");
            yield return new DesktopModelCompleted(
                "toolUse",
                new DesktopModelUsage(
                    Input: 12,
                    Output: 8,
                    CacheRead: 0,
                    CacheWrite: 0));
            yield break;
        }
        if (sequence == 4)
        {
            yield return new DesktopModelTextDelta(
                "JARVIS workspace tool online.");
            yield return new DesktopModelCompleted(
                "stop",
                new DesktopModelUsage(
                    Input: 18,
                    Output: 5,
                    CacheRead: 0,
                    CacheWrite: 0));
            yield break;
        }

        yield return new DesktopModelTextDelta("JARVIS ");
        yield return new DesktopModelTextDelta(
            "desktop broker online.");
        yield return new DesktopModelCompleted(
            "stop",
            new DesktopModelUsage(
                Input: 7,
                Output: 4,
                CacheRead: 0,
                CacheWrite: 0));
    }
}
